src/jennifer/config.cr

```diff
@@ -4,139 +4,179 @@ require "logger"
+        @retry_attempts = 1
+
+        @checkout_timeout = 5.0
+        @retry_delay = 1.0
+
+        @logger = Logger.new(STDOUT)
+        @logger.not_nil!.level = Logger::DEBUG
+        @logger.not_nil!.formatter = Logger::Formatter.new do |severity, datetime, progname, message, io|
+          io << datetime << ": " << message
+        end
       end
-    {% end %}
-    end
 
-    define_fields(STRING_FIELDS, default: "")
-    define_fields(INT_FIELDS, 0)
-    define_fields(FLOAT_FIELDS, 0.0)
+      def structure_folder
+        if @structure_folder.empty?
+          File.dirname(@migration_files_path)
+        else
+          @structure_folder
+        end
+      end
 
-    def self.structure_folder
-      if @@structure_folder.empty?
-        File.dirname(@@migration_files_path)
-      else
-        @@structure_folder
+      def structure_path
+        File.join(self.structure_folder, "structure.sql")
       end
-    end
 
-    def self.structure_path
-      File.join(Config.structure_folder, "structure.sql")
-    end
+      def logger
+        @logger.not_nil!
+      end
 
-    def self.reset_config
-      @@adapter = "postgres"
-      @@host = "localhost"
-      @@port = -1
-      @@migration_files_path = "./db/migrations"
-      @@schema = "public"
-      @@db = ""
-
-      @@initial_pool_size = 1
-      @@max_pool_size = 5
-      @@max_idle_pool_size = 1
-      @@retry_attempts = 1
-
-      @@checkout_timeout = 5.0
-      @@retry_delay = 1.0
-
-      @@logger = Logger.new(STDOUT)
-      @@logger.not_nil!.level = Logger::DEBUG
-      @@logger.not_nil!.formatter = Logger::Formatter.new do |severity, datetime, progname, message, io|
-        io << datetime << ": " << message
+      def logger=(value)
+        @logger = value
       end
-    end
 
-    reset_config
+      def validate_config
+        raise Jennifer::InvalidConfig.new("No adapter configured") if adapter.empty?
+        raise Jennifer::InvalidConfig.new("No database configured") if db.empty?
+      end
 
-    def self.logger
-      @@logger.not_nil!
-    end
+      def connection_string(*options)
+        auth_part = @user
+        auth_part += ":#{@password}" if @password && !@password.empty?
+
+        host_part = @host
+        host_part += ":#{@port}" if @port && @port > 0
+
+        String.build do |s|
+          s << @adapter << "://"
+          s << auth_part << "@" if auth_part.size > 0
+          s << host_part
+          s << "/" << @db if options.includes?(:db)
+          s << "?"
+          {% for arg, index in CONNECTION_URI_PARAMS %}
+            s << "{{arg.id}}=#{@{{arg.id}}}"
+            s << "&" if {{index}} < {{CONNECTION_URI_PARAMS.size - 1}}
+          {% end %}
+        end
+      end
 
-    def self.logger=(value)
-      @@logger = value
-    end
+      def from_uri(db_uri : String)
+        begin
+          from_uri(URI.parse(db_uri))
+        rescue e : URI::Error
+          self.logger.error("Error parsing database uri #{db_uri} - #{e.message}")
```

Comment:
hm, I'm not sure we should prevent raising exception of broken db uri

Reply:
I'll cut a separate pr to reraise and test this.